services/navigation/builtin/builtin.go

```diff
@@ -419,16 +459,58 @@ func (svc *builtIn) startWaypointExperimental(extra map[string]interface{}) erro
 		}
 
 		// loop until no waypoints remaining
-		for wp, err := svc.nextWaypoint(svc.cancelCtx); err == nil; wp, err = svc.nextWaypoint(svc.cancelCtx) {
+		for wp, err := svc.nextWaypoint(cancelCtx); err == nil; wp, err = svc.nextWaypoint(cancelCtx) {
+			if cancelCtx.Err() != nil {
+				return
+			}
+
+			navOnceCancelCtx := svc.resetForNavOnce(cancelCtx, wp)
 			svc.logger.Infof("navigating to waypoint: %+v", wp)
-			if err := navOnce(svc.cancelCtx, wp); err != nil {
+			if err := navOnce(navOnceCancelCtx, wp); err != nil {
+				if svc.waypointIsDeleted() {
+					svc.logger.Infof("skipping waypoint %+v since it was deleted", wp)
+					continue
+				}
+
 				svc.logger.Infof("skipping waypoint %+v due to error while navigating towards it: %s", wp, err)
+				if err := svc.waypointReached(cancelCtx); err != nil {
+					svc.logger.Info("can't mark waypoint %+v as reached, exiting navigation due to error: %s", wp, err)
+				}
 			}
 		}
 	})
-	return nil
+}
+
+func (svc *builtIn) resetForNavOnce(ctx context.Context, wp navigation.Waypoint) context.Context {
```

Comment:
Lets not add this as a helper function, it looks like its only being used once so I think we can just put it in startWaypointExperimental